### PR TITLE
Catch symdiff errors.

### LINKED ISSUE #1893: modcc segfaults in conjunction with comparison
Hi!

I'm on `bb723be670f3b5a85da8b40d0e41fe0a18f240b5` and have the following start for the implementation of a Graupner/Brunel mechanism:

```
NEURON {
    POINT_PROCESS graupner_brunel
}

INITIAL {
    c = 0
    rho = 0
    theta_p = 0
}

STATE {
    c
    rho
    theta_p
}

PARAMETER {
    tau_c = 150 (ms)
}

BREAKPOINT {
    SOLVE state METHOD cnexp
}

DERIVATIVE state {
    c' = -c/tau_c
    rho' = (c - theta_p) > 0
}
```

This segfaults with:

```
make_unique_local_assign (scope=std::shared_ptr<Scope<Symbol>> (use count 24, weak count 0) = {...}, e=0x0, prefix="b_") at /home/sschmitt/HBP/arbor/modcc/astmanip.cpp:16
16          Location loc = e->location();
(gdb) where
#0  make_unique_local_assign (scope=std::shared_ptr<Scope<Symbol>> (use count 24, weak count 0) = {...}, e=0x0, prefix="b_") at /home/sschmitt/HBP/arbor/modcc/astmanip.cpp:16
#1  0x00005555555cfb59 in CnexpSolverVisitor::visit (this=0x55555573d3a0, e=0x555555745400) at /home/sschmitt/HBP/arbor/modcc/solvers.cpp:56
#2  0x0000555555683c62 in AssignmentExpression::accept (this=0x555555745400, v=0x55555573d3a0) at /home/sschmitt/HBP/arbor/modcc/expression.cpp:1102
#3  0x00005555555a3d6c in BlockRewriterBase::visit (this=0x55555573d3a0, e=0x555555745490) at /home/sschmitt/HBP/arbor/modcc/visitor.hpp:114
#4  0x00005555555cf83b in CnexpSolverVisitor::visit (this=0x55555573d3a0, e=0x555555745490) at /home/sschmitt/HBP/arbor/modcc/solvers.cpp:27
#5  0x000055555568361c in BlockExpression::accept (this=0x555555745490, v=0x55555573d3a0) at /home/sschmitt/HBP/arbor/modcc/expression.cpp:1015
#6  0x000055555559c7eb in Module::semantic (this=0x7fffffffdb00) at /home/sschmitt/HBP/arbor/modcc/module.cpp:440
#7  0x0000555555576ccd in main (argc=1, argv=0x7fffffffe118) at /home/sschmitt/HBP/arbor/modcc/modcc.cpp:224
```

Removing the comparison

```
	rho' = c - theta_p
```

gives

```
error: error   graupner_brunel.mod:(line 27,col 7) System not diagonal linear for cnexp
```

Changing to sparse

```
    SOLVE state METHOD sparse
```

gives

```
error: error   graupner_brunel.mod:(line 27,col 7) System not homogeneous linear for sparse
```

Thanks,

Sebastian

### PR DESCRIPTION
**Pull request by thorstenhater**
- Make symbolic differentiation return a more refined type.
- Catch said error when trying to test for linearity
- Catch the propagated error in CNExp and recommend using a different one.
- Straighten out the error printing at top level
- Add regression tests

Closes #1893

New output show the full error stack
```
bin/modcc bug_1893/bug_1893.mod -N 'arb::allen_catalogue' -t cpu -t gpu
error trace:
  * bug_1893/bug_1893.mod:(line 27,col 26) symbolic differential of unrecognized binary expression
  * bug_1893/bug_1893.mod:(line 27,col 10) CNExp: Could not determine linearity, maybe use a different solver?
```

### CODE REVIEW

File: modcc/symdiff.cpp [hunk 1]
```diff
@@ -666,17 +676,20 @@ ARB_LIBMODCC_API linear_test_result linear_test(Expression* e, const std::vector
...
+        if (!coef) return linear_test_result{};
         if (!is_zero(coef)) result.coef[id] = std::move(coef);
-
         result.constant = substitute(result.constant, id, zero());
     }
 
     ConstantSimplifyVisitor csimp_visitor;
     result.constant->accept(&csimp_visitor);
     result.constant = csimp_visitor.result();
+    if (result.constant.get() == nullptr) throw compiler_exception{"Simplify: linear has no constant term.", loc};
```
Comment (boeschf) [inline modcc/symdiff.cpp:692]: why is it a problem if there is no constant term?

Comment (thorstenhater) [reply modcc/symdiff.cpp:692]: Oh, that's not really the core issue. The `NULL` value is used to signal an error by the visitor. 
While the visitor has an internal error state that's pretty elaborate, it does not bubble that up 
to the user. 
 
The message could be better, though.

Comment (boeschf) [reply modcc/symdiff.cpp:692]: gotcha, yes I think the error message should reflect that, and not state that there is no constant term

**Review by boeschf — DISMISSED** [review]
Looks good, but please consider changing the error message

**Comment by thorstenhater** [comment]
@boeschf I fear you to r'approve.

**Review by boeschf — APPROVED** [review]

**Pull request merged**